[dtensor] tensor ops to use strategy based sharding prop

wanchaol opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #100607

This is the first series of PR that adopts operator impls to use a
strategy based approach, each op utilizes OpStrategy and PlacementStrategy
to generate their own strategy. By utilizing the strategy based
approach along with the op graph, we could enable more advanced op
implementation (decomp is possible), and turn the sharding prop to be
more like a contraint satisfication problem.

This PR alone only adds some basic tensor op strategies, and it directly
works on the op graph that was used for metadata propagation. The tensor ops
added in this PR mainly follows one of the arg strategy. The next set of
PRs would add more op strategies to other ops.

wanchaol committed:
[dtensor] tensor ops to use strategy based sharding prop

This is the first series of PR that adopts operator impls to use a
strategy based approach, each op utilizes OpStrategy and PlacementStrategy
to generate their own strategy. By utilizing the strategy based
approach along with the op graph, we could enable more advanced op
implementation (decomp is possible), and turn the sharding prop to be
more like a contraint satisfication problem.

This PR alone only adds some basic tensor op strategies, and it directly
works on the op graph that was used for metadata propagation. The tensor ops
added in this PR mainly follows one of the arg strategy. The next set of
PRs would add more op strategies to other ops.

[ghstack-poisoned]

XilunWu commented on torch/distributed/_tensor/sharding_prop.py:
```diff
@@ -54,14 +76,114 @@ def prepare_op_schema(
+                self._wrap_output_spec_meta(output_sharding.output_spec, output_node)
+            return output_sharding
+
+        elif self.op_to_rules.get(op_overload, None) is not None:
```
see above

XilunWu reviewed (commented):
sorry for the late review! Some questions and suggestions. I have a question on the high-level side, the new change utilizes fx graph, does that mean DTensor will eventually move from eager execution to graph execution (i.e. compiler mode)?

wanchaol committed:
Update on "[dtensor] tensor ops to use strategy based sharding prop"


This is the first series of PR that adopts operator impls to use a
strategy based approach, each op utilizes OpStrategy and PlacementStrategy
to generate their own strategy. By utilizing the strategy based
approach along with the op graph, we could enable more advanced op
implementation (decomp is possible), and turn the sharding prop to be
more like a contraint satisfication problem.

This PR alone only adds some basic tensor op strategies, and it directly
works on the op graph that was used for metadata propagation. The tensor ops
added in this PR mainly follows one of the arg strategy. The next set of
PRs would add more op strategies to other ops.

[ghstack-poisoned]

wanchaol committed:
Update on "[dtensor] tensor ops to use strategy based sharding prop"


This is the first series of PR that adopts operator impls to use a
strategy based approach, each op utilizes OpStrategy and PlacementStrategy
to generate their own strategy. By utilizing the strategy based
approach along with the op graph, we could enable more advanced op
implementation (decomp is possible), and turn the sharding prop to be
more like a contraint satisfication problem.

This PR alone only adds some basic tensor op strategies, and it directly
works on the op graph that was used for metadata propagation. The tensor ops
added in this PR mainly follows one of the arg strategy. The next set of
PRs would add more op strategies to other ops.

[ghstack-poisoned]

wanchaol committed:
Update on "[dtensor] tensor ops to use strategy based sharding prop"


This is the first series of PR that adopts operator impls to use a
strategy based approach, each op utilizes OpStrategy and PlacementStrategy
to generate their own strategy. By utilizing the strategy based
approach along with the op graph, we could enable more advanced op
implementation (decomp is possible), and turn the sharding prop to be
more like a contraint satisfication problem.

This PR alone only adds some basic tensor op strategies, and it directly
works on the op graph that was used for metadata propagation. The tensor ops
added in this PR mainly follows one of the arg strategy. The next set of
PRs would add more op strategies to other ops.

[ghstack-poisoned]

wanchaol commented:
> does that mean DTensor will eventually move from eager execution to graph execution (i.e. compiler mode)?

We are already using the small op graph for metadata propagation (i.e. output shape/stride), so I used it for strategy based sharding prop too, I don't really know whether we should use this op graph for runtime execution yet. I feel we should keep using the eager execution and only do sharding prop on the graph, if we later found it might be good to directly use the op graph to run, we should evaluate the perf and switch to use that for execution afterwards.

wanchaol committed:
Update on "[dtensor] tensor ops to use strategy based sharding prop"


This is the first series of PR that adopts operator impls to use a
strategy based approach, each op utilizes OpStrategy and PlacementStrategy
to generate their own strategy. By utilizing the strategy based
approach along with the op graph, we could enable more advanced op
implementation (decomp is possible), and turn the sharding prop to be
more like a contraint satisfication problem.

This PR alone only adds some basic tensor op strategies, and it directly
works on the op graph that was used for metadata propagation. The tensor ops
added in this PR mainly follows one of the arg strategy. The next set of
PRs would add more op strategies to other ops.

[ghstack-poisoned]

XilunWu reviewed (approved):
lgtm! Thanks for adding strategy-based sharding propagation that may simplify implementing new tensor ops.

wanchaol commented:
@pytorchbot merge

wanchaol commented:
@pytorchbot merge

wanchaol commented:
@pytorchbot merge

wanchaol commented:
@pytorchbot merge

wanchaol commented:
@pytorchbot merge

wanchaol commented:
@pytorchbot rebase

wanchaol committed:
Update on "[dtensor] tensor ops to use strategy based sharding prop"


This is the first series of PR that adopts operator impls to use a
strategy based approach, each op utilizes OpStrategy and PlacementStrategy
to generate their own strategy. By utilizing the strategy based
approach along with the op graph, we could enable more advanced op
implementation (decomp is possible), and turn the sharding prop to be
more like a contraint satisfication problem.

This PR alone only adds some basic tensor op strategies, and it directly
works on the op graph that was used for metadata propagation. The tensor ops
added in this PR mainly follows one of the arg strategy. The next set of
PRs would add more op strategies to other ops.

[ghstack-poisoned]

wanchaol commented:
@pytorchbot merge
